### ISSUE TITLE
Widely used names like Event or Link and API design bests practices

### ISSUE THREAD

**Issue by lmolkova** [open-telemetry/opentelemetry-specification#318]
We define many types with short and very common names like `Link`, `Event`, even `Span` (e.g. Span in .NET is something different and it conflicts with OTel span, `event` is keyword in C#)

Chances are it conflicts with something in specific language/common framework or library.

Now imagine that `Event` is used by some library. Now users add OTel and get `Event` from library and from OTel. Is it easy to write code like this? No.

Imagine that the library owner decides to instrument with OTel. It's pain for them and pain for all users of their library (even if they don't use OTel).

There are some recommendations to avoid short, ambiguos, common names.

- [Google naming conventions](https://cloud.google.com/apis/design/naming_convention
) from Google to avoid this: 
 
> Avoid overly general names that are ambiguous within the context of the API and the larger ecosystem of Google APIs. They can lead to misunderstanding of API concepts. Rather, choose specific names that accurately describe the API concept. This is particularly important for names that define first-order API elements, such as resources. There is no definitive list of names to avoid, as every name must be evaluated in the context of other names. Instance, info, and service are examples of names that have been problematic in the past. Names chosen should describe the API concept clearly (for example: instance of what?) and distinguish it from other relevant concepts (for example: does "alert" mean the rule, the signal, or the notification?).

- There are similar principles for .NET API design (could not find link) that .NET core team shared with us reviewing OTel APIs.

Googling for API design naming brings some more results like 

- [Python naming conventions](https://visualgit.readthedocs.io/en/latest/pages/naming_convention.html#python-naming-conventions)

> Avoid using names that are too general or too wordy. Strike a good balance between the two.

- [The Little Manual of API Design by J Blanchette - ‎2008 - ‎Cited by 29](https://people.mpi-inf.mpg.de/~jblanche/api-design.pdf#page=25&zoom=100,0,814)

> Prefer specific names to general ones, even if they sometimes seem too restrictive. Remember that names occupy API real estate: Once a name is
taken, it cannot be used for something else. This is especially an issue with
class names, since a library might contain hundreds or even thousands of
classes that need distinct names. Another reason for preferring more specific names is that they are easier for users to relate to.

Considering all this, can we look for a bit better naming scheme that would not conflict with extremely widely used terms and does not violate typical recommendations?

**Comment by rektide** [comment]
I don't feel like putting "OpenTelementry" in front of every entity helps anything. Every programming language you've mentioned so far has some kind of namespacing. I'm not sure why we'd alter the spec to make `Event` into `OpenTelemetryEvent` or `Link` into `OpenTelemetryLink` &c.

That's what I infer you asking for, at least. Do you have alternative ideas for what a "better naming scheme" might consist of?

**Comment by lmolkova** [comment]
I'm not suggesting prefixing with OpenTelemtry. Before coming up with proposal I want to understand how the community thinks about it and if other members consider it as an issue.

**Comment by lmolkova** [comment]
BTW, namespacing is not very helpful here, if you need to work on namespaceA.Link and namespaceB.Link in the same file

**Comment by tigrannajaryan** [comment]
@lmolkova I agree with your argumentation and links you posted are very useful (and it is not really about namespacing but about using names that are properly specific). However it may be too late now to rename these broadly used concepts since it usually means a lot of work across large number of repos and SIGs (I know, I had to rename Service to Collector a couple weeks ago and it was a LOT of work just for single name change). I am not against renaming but want to point out that ROI of the renaming effort may be negative.

**Comment by lmolkova** [comment]
I agree it involves a lot of work and wish we've got this feedback much earlier. Renaming today is very expensive. It's also common terminology people in tracing used for years. But today this is still possible but after v1 it won't be, ever. 

The reason I'm still bringing it is that I hope more integrations with all kinds of libraries and frameworks that are done and shipped by library owners rather than OTel community. And I boldly think I influence some of them among Microsoft libraries. I believe some API choices like this, can hurt this adoption in the future.

**Comment by jmacd** [comment]
Before adding my support to #514, I'll enter an opinion here from the Golang API perspective. Go uses a package prefix whenever referring to an external type or function name, so we do not refer to a `Span` directly, we refer to `trace.Span`, for example. I feel that each language should use the name-spacing provided in that language, so that users are not referring to widely-used names in their program. This lets us use a term like `Span` in our specification without occupying such a large "API real estate" when used in the language.

**Comment by andrewhsu** [comment]
from the spec sig mtg today, @SergeyKanzhelev would this issue still need to be address?

**Comment by carlosalberto** [comment]
Ping @SergeyKanzhelev @reyang - is this still an issue for the C#/.Net land?

**Comment by reyang** [comment]
I don't think so, happy to close it. In general each language should follow their idiomatic way and we can have a general wording in the spec.
